### DISCUSSION TITLE
Dependencies

### DISCUSSION THREAD

**Discussion by ktaeyln** [community/community#9464] [Projects and Issues]
If I missed it I'm sorry, but, I'm not seeing a way to make issues/milestones and project goals depend on other issues/milestones and project goals... which makes it hard to see this as a viable project management system...

**Comment by ThierryA** [comment]
+1 to have the ability to manage issue dependency (a dependency or issue linking field as part of the new project table). ZenHub does that quite well and I would love to see that as a native GitHub feature!

**Comment by Jille** [comment]
You can use [SnoozeThis](https://github.com/apps/snoozethis) to automatically close your epics when all their children closes. While issues are blocked, they'd be assigned to SnoozeThis. (Additionally, you can also automatically add/remove a "blocked" label.)

Simply comment this on your epic:
```
@SnoozeThis <url of first child>, <url of second child> -> closed
```

**Comment by turadg** [comment]
https://github.com/features/issues#faq says,
> With GitHub Issues, … you can break big issues into tasks, further organize your work with milestones and labels, and **track relationships and dependencies**.

Is this feature planned? Or is the text just about the relationship of an issue with a PR?

**Comment by rileybroughten** [comment]
👋 Posing some questions for some more feedback regarding item dependencies!
- What do you most often use dependencies for? 
- Would dependencies be helpful more as a reference on an issue or in a project that you can use to gather more information, or would you expect certain behaviors to occur automatically between your dependent items?
- Do you think of "depends on" and "blocked by" as the same thing or different?

Curious to hear more thoughts on this 😁

**Reply by ktaeyln** [reply]
- More complex hierarchical issues/milestones
- So with milestones, having one milestone depend on another would mean it would inherit the issues from the one it depends on. With issues:
- Blocked by would be the same thing, if one issue depends on another it would block completion until the one it depends on is also completed

It could just be referencial as well, but it would be a visual way to help say: "This issue is more or a priority because hundreds of other issues depend on it", it would also be nice to be able to close an issue that is a duplicate by saying it "depends" on the original in such a way that it outright lists the duplicates in the dependencies, to demonstrate visually and numerically how many times this issue has been reported; obviously if people are reporting the same issue 15000 times it should have more priority... hence sorting by issues that have the most dependants would become super useful, potentially.

**Reply by mike-engel** [reply]
- Tracking issues as part of a larger feature. We break features down into smaller issues that can be developed, tested, and shipped with a feature flag constantly until the feature is done. Some parts of these features can't be started until an earlier issue is finished.
- I would expect some sort of visual flag on a project board/issue which would change based on dependency statuses to indicate that:
  - An issue is blocking other issues, and which ones
  - An issue is being blocked by other issues, and which ones
- For our team, they're the same thing

For now, we're manually managing this visual flag using labels for "Blocked" (red) and "Blocker" (brown/yellow). On a project board it's then easy to see which issues can't be worked on yet, and which ones should be worked on immediately to remove the dependency/blocker. Inside the issue we link to the dependencies which helps, but changing the labels and monitoring the links in the issue is a manual process and error prone.

**Reply by siennathesane** [reply]
Hi @rileybroughten, here's another data point

* _What do you most often use dependencies for?_
  * To track complex relationships between issues, pull requests, or other requirements/actions
* _Would dependencies be helpful more as a reference on an issue or in a project that you can use to gather more information, or would you expect certain behaviors to occur automatically between your dependent items?_
  * If the relationship can be modeled as a reference, then built-in actions can come over time.
* _Do you think of "depends on" and "blocked by" as the same thing or different?_
  * `depends on` would be like saying, _my life depends on breathing oxygen_. `blocked by` would be like saying _breathing oxygen is blocked by something in my airway_. They're two different `typesOf` relationships that can model transitive dependencies or different actions.

Hopefully that helps!

**Reply by mxbclang** [reply]
Hi @rileybroughten! Our team would love to switch over to GH Projects but dependencies are a major blocker for us. To answer your questions:

**What do you most often use dependencies for?**
We are often building complex features with at least 30+ issues that have multiple dependencies. For example, here's a screenshot of a dependency map for a recent feature: 

![image](https://github.com/community/community/assets/12189958/3ba74d1b-663f-4683-9bee-93e1c7dcbbb3)

Within GitHub, we need to be able to mark these issues accordingly. As @ThierryA noted above, ZenHub (which we use now) [does this quite well](https://help.zenhub.com/support/solutions/articles/43000010349-creating-github-issue-dependencies-in-zenhub). Per the image above, we're able to mark 6256, for example as **dependent on** 6308 and **blocking** 6257. 

**Would dependencies be helpful more as a reference on an issue or in a project that you can use to gather more information, or would you expect certain behaviors to occur automatically between your dependent items?**

The only behavior I'd expect is something similar to what ZenHub does, where there's are indicators on the issues that are dependent on/blocking. These appear both within the issue itself as well as on the project board:

Blocked issue:
![image](https://github.com/community/community/assets/12189958/1784aae9-b9c0-4797-a1b4-da8bb708e95f)

![image](https://github.com/community/community/assets/12189958/11535daf-c645-45a7-9d15-4fc9c996c63c)

Dependent issue:

![image](https://github.com/community/community/assets/12189958/31151d3d-22d0-4a52-a7cb-15980ac4189b)

![image](https://github.com/community/community/assets/12189958/d01532e4-4e64-474e-8215-ce04a22902b2)

**Do you think of "depends on" and "blocked by" as the same thing or different?**

I'd say they're basically the same to us – both indicate that something cannot move without the completion of something else.

I hope this is helpful! Please feel free to get in touch with any other questions – we'd really love to see this feature added to GH Projects. Thanks!

**Reply by JonB451** [reply]
Thanks @bethanylang 

Is this DAG produced in ZenHub too, or is that your own visualization?

**Reply by mxbclang** [reply]
@JonB451 My own low-res one in a Google drawing.  :)

**Reply by JonB451** [reply]
@bethanylang  We'll take low-res functionality over high-res non-functionality any day :) Thanks.

**Comment by JonB451** [comment]
Thanks for raising this @ktaeyln 

Is there a reason that no one seems to mention a DAG visualization of task dependencies, or even a DAG interface? That is what we are really looking for.

**Reply by nexushoratio** [reply]
A few days ago I'd realized my issue graph was no longer acyclic.  Oops.

**Comment by RumeysaaKopuz** [comment]
response

**Comment by bkuberek** [comment]
Similarly, dependencies between pull requests.

I would like to see the ability to block pull request merges based on other pull requests that are open, creating a dependency between two or more pull requests. 

For example, project A has a library that needs to be updated before project B can merge the service that uses such library. In this scenario, Project B / pull request B1, has a dependency on project A / pull request A1. A1 blocks B1 from being merged until A1 is merged.

I'm not sure if this has already been mentioned before. It's just a thought.
